VariantShredding.md

```diff
@@ -25,276 +25,302 @@
+|----------------------------------|-------------|---------------|-----------------------|--------------------------------|
+| `["comedy", "drama"]`            | null        | non-null      | [null, null]          | [`comedy`, `drama`]            |
+| `["horror", null]`               | null        | non-null      | [null, `00`]          | [`horror`, null]               |
+| `["comedy", "drama", "romance"]` | null        | non-null      | [null, null, null]    | [`comedy`, `drama`, `romance`] |
+| null                             | `00` (null) | null          |                       |                                |
 
-Reconstruction of Variant value from a shredded representation is not expected to produce a bit-for-bit identical binary to the original unshredded value.
-For example, the order of fields in the binary may change, as may the physical representation of scalar values.
+#### Objects
 
-In particular, the [Variant Binary Encoding](VariantEncoding.md) considers all integer and decimal representations to represent a single logical type.
-As a result, it is valid to shred a decimal into a decimal column with a different scale, or to shred an integer as a decimal, as long as no numeric precision is lost.
-For example, it would be valid to write the value 123 to a Decimal(9, 2) column, but the value 1.234 would need to be written to the **variant_value** column.
-When reconstructing, it would be valid for a reader to reconstruct 123 as an integer, or as a Decimal(9, 2).
-Engines should not depend on the physical type of a Variant value, only the logical type.
+Fields of an object can be shredded using a Parquet group for `typed_value` that contains shredded fields.
 
-On the other hand, shredding as a different logical type is not allowed.
-For example, the integer value 123 could not be shredded to a string `typed_value` column as the string "123", since that would lose type information.
-It would need to be written to the `variant_value` column.
+If the value is not an object, `typed_value` must be null.
 
-# Reconstructing a Variant
+If the value is a partially shredded object, the `value` must not contain the shredded fields.
+If shredded fields are present in the variant object, it is invalid and readers must either fail or use the shredded values.
 
-It is possible to recover a full Variant value using a recursive algorithm, where the initial call is to `ConstructVariant` with the top-level fields, which are assumed to be null if they are not present in the schema.
+Each shredded field in the `typed_value` group is represented as a required group that contains optional `value` and `typed_value` fields.
+The `value` field stores the value as Variant-encoded `binary` when the `typed_value` cannot represent the field.
+This layout enables readers to skip data based on the field statistics for `value` and `typed_value`.
```

Comment:
IIUC, field statistics for `value` does not add too much value except for all nulls?

Reply:
The lower/upper bounds can be used to detect when all values are variant null.

Reply:
This brings up an interesting use case.  It seems for value data having the type information in the most significant bits would allow further conferences like this (e.g. is it always a strong)

Reply:
I think this is something we should discuss more widely.  Variant is still experimental from Parquet's perspective, and the binary metadata supports versioning.

I think there are a couple of ways we can make this work and not break existing variants in the wild (maybe there are others):
1.  Disambiguate on the logical type annotation which hasn't been released yet.
2. Use the metadata version to track the version 

I think option 2 is probably the most viable.

Reply:
Another option is to add a statistics field to track which types are present in the value, this is more expensive then then min/max for what I believe the common case of likely only having one alternate type.

Reply:
I replied on the dev list, but I don't think this is worth making a breaking change to the encoding. There's little value here, especially if the non-shredded values end up being dictionary-encoded.

Reply:
I'm OK with this, just wanted to make sure this was discussed it seems no one else has raised concerns on the dev list.